packages/angular_devkit/build_angular/src/webpack/plugins/occurrences-plugin.ts

```diff
@@ -0,0 +1,84 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import { Compiler } from 'webpack';
+
+const PLUGIN_NAME = 'angular-occurrences-plugin';
+
+export class OccurrencesPlugin {
+  apply(compiler: Compiler) {
+    compiler.hooks.thisCompilation.tap(PLUGIN_NAME, (compilation) => {
+      compilation.hooks.processAssets.tapPromise(
+        {
+          name: PLUGIN_NAME,
+          stage: compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_ANALYSE,
+        },
+        async (compilationAssets) => {
+          for (const assetName of Object.keys(compilationAssets)) {
+            if (!assetName.endsWith('.js')) {
+              continue;
+            }
+
+            const scriptAsset = compilation.getAsset(assetName);
+            if (!scriptAsset || scriptAsset.source.size() <= 0) {
+              continue;
+            }
+
+            const src = scriptAsset.source.source().toString('utf-8');
+
+            let ngComponentCount = 0;
+            // Count the number of `Component({` strings (case sensitive), which happens in __decorate().
+            ngComponentCount += this.countOccurrences(src, 'Component({');
+            // For Ivy we just count ɵcmp.src
+            ngComponentCount += this.countOccurrences(src, '.ɵcmp', true);
+            // for ascii_only true
+            ngComponentCount += this.countOccurrences(src, '.\\u0275cmp', false);
```

Comment:
Could we potentially optimize this? In AOT mode, i don't think there would be any `Component({` decorators.  Also, when optimization is enabled then `.ɵcmp` shouldn't exist in the output.

Reply:
Yes correct. Let's do that.